File: frontend/src/api/twin.ts
export type TwinRunState = "queued" | "running" | "completed" | "cancelled" | "failed";

export interface ScenarioDefinition {
  code: string;
  name: string;
  category: string;
  description: string;
  parameters: Record<string, unknown>;
  explanation: string;
  scenario_version: string;
  source_type: "internal_demo";
  is_composable: boolean;
  enabled: boolean;
}

export interface ScenarioCatalog {
  scenario_version: string;
  source_type: "internal_demo";
  source_summary: string;
  scenario_count: number;
  scenarios: ScenarioDefinition[];
}

export interface TwinRunRequest {
  analysis_date: string;
  seed: number;
  path_count: number;
  horizon_years: number;
  output_interval_months: 1 | 3 | 6 | 12;
  scenario_codes: string[];
  scenario_overrides: {
    unemployment_months?: number;
    income_reduction_ratio?: string;
    medical_shock_amount?: string;
    education_overrun_amount?: string;
    property_value_change_ratio?: string;
    mortgage_rate_change?: string;
  };
  plan_adjustments: {
    primary_retirement_age: number;
    additional_monthly_savings: string;
    equity_ratio: string;
    liquidity_reallocation_amount: string;
  };
  assumption_overrides: {
    inflation_rate?: string;
    income_growth_rate?: string;
  };
  family_events: Array<{
    code: string;
    name: string;
    start_month: number;
    duration_months: number;
    one_time_cost: string;
    monthly_income_loss: string;
    monthly_expense_increase: string;
  }>;
}

export interface FanPoint {
  month: number;
  date: string;
  primary_age: string;
  p10: string;
  p25: string;
  p50: string;
  p75: string;
  p90: string;
  median_liquid_assets: string;
  median_long_term_assets: string;
  median_liabilities: string;
}

export interface GoalSimulationOutcome {
  goal_id: string;
  name: string;
  due_month: number;
  required_amount: string;
  success_probability: string;
  failure_probability: string;
  median_shortfall: string;
}

export interface SimulationDistribution {
  label: string;
  path_count: number;
  horizon_months: number;
  goal_success_probability: string;
  depletion_probability: string;
  forced_sale_probability: string;
  median_forced_sale_amount: string;
  ending_net_worth_median: string;
  ending_net_worth_p10: string;
  total_goal_shortfall_median: string;
  required_additional_monthly_savings: string;
  fan: FanPoint[];
  goal_outcomes: GoalSimulationOutcome[];
  failure_time_distribution: Array<{
    year: number;
    path_count: number;
    probability: string;
    most_common_goal: string | null;
  }>;
  worst_paths: Array<{
    path_id: number;
    ending_net_worth: string;
    minimum_net_worth: string;
    first_failed_goal: string | null;
    first_failure_month: number | null;
    depletion_month: number | null;
    forced_sale_amount: string;
    total_goal_shortfall: string;
    explanation: string;
  }>;
  validation: {
    all_values_finite: boolean;
    primary_income_paid_during_interruption_max: string;
    goal_spending_events_applied: number;
    common_random_numbers: boolean;
  };
}

export interface ScenarioImpact {
  scenario_code: string;
  name: string;
  emergency_support_months: string;
  first_failed_goal: string | null;
  forced_sale_probability: string;
  median_forced_sale_amount: string;
  insurance_coverage_applied: string;
  remaining_medical_gap: string;
  retirement_delay_needed: boolean;
  suggested_retirement_delay_years: number;
  monthly_compressible_expenses: string;
  required_additional_monthly_savings: string;
  baseline_goal_success_probability: string;
  scenario_goal_success_probability: string;
  success_probability_change: string;
  applicable: boolean;
  explanation: string;
}

export interface TwinResult {
  meta: {
    run_id: string;
    household_id: string;
    household_code: string;
    analysis_date: string;
    data_as_of: string;
    input_version: string;
    rule_code: string;
    rule_version: string;
    formula_version: string;
    engine_version: string;
    result_version: string;
    scenario_version: string;
    calculation_source: "deterministic_simulation_engine";
    synthetic_data: boolean;
    currency: string;
  };
  initial_state: {
    members: Array<{
      member_id: string;
      name: string;
      relationship: string;
      age_at_start: string;
      recorded_retirement_age: number | null;
    }>;
    annual_income: string;
    annual_expenses_excluding_debt_service: string;
    annual_essential_expenses_excluding_debt_service: string;
    monthly_compressible_expenses: string;
    asset_buckets: Record<string, string>;
    total_assets: string;
    total_liabilities: string;
    medical_coverage_available: string;
    medical_deductible: string;
    pension_annual_contributions: string;
    goals: Array<{
      goal_id: string;
      name: string;
      goal_type: string;
      due_month: number;
      current_amount: string;
      prepared_amount: string;
      annual_cost_growth_rate: string;
    }>;
    source_record_ids: string[];
    counting_note: string;
  };
  assumptions: {
    seed: number;
    path_count: number;
    horizon_months: number;
    time_step_months: number;
    output_interval_months: number;
    inflation_rate: string;
    income_growth_rate: string;
    asset_classes: string[];
    asset_assumptions: Record<string, { expected_annual_return: string; annual_volatility: string }>;
    correlation_matrix: string[][];
    property_assumption: { expected_annual_return: string; annual_volatility: string };
    pension_assumption: { expected_annual_return: string; annual_volatility: string };
    scenario_codes: string[];
    scenario_parameters: Record<string, unknown>;
    plan_adjustments: TwinRunRequest["plan_adjustments"];
    family_events: TwinRunRequest["family_events"];
    parameter_hash: string;
    source_type: "internal_demo";
  };
  baseline: SimulationDistribution;
  original_stress: SimulationDistribution;
  optimized_stress: SimulationDistribution;
  scenario_impacts: ScenarioImpact[];
  comparison: {
    baseline_success_probability: string;
    original_stress_success_probability: string;
    optimized_stress_success_probability: string;
    stress_change: string;
    optimization_change: string;
    baseline_forced_sale_probability: string;
    original_forced_sale_probability: string;
    optimized_forced_sale_probability: string;
    liquidity_reallocation_amount: string;
    avoided_forced_sale_probability: string;
    stress_not_better_than_baseline: boolean;
    positive_override_explanation: string | null;
    liquidity_explanation: string;
  };
  limitations: string[];
  counting_note: string;
}

export interface TwinRunStatus {
  run_id: string;
  household_id: string;
  status: TwinRunState;
  progress_percent: number;
  phase: string;
  scenario_codes: string[];
  path_count: number;
  horizon_months: number;
  seed: number;
  input_version: string;
  rule_version: string;
  engine_version: string;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  cancel_requested: boolean;
  audit_event_id: string | null;
  error_code: string | null;
  result: TwinResult | null;
}

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? "").replace(/\/$/, "");

async function readJson<T>(path: string, signal?: AbortSignal, init?: RequestInit): Promise<T> {
  const response = await fetch(apiBase + path, {
    ...init,
    signal,
    headers: { Accept: "application/json", ...init?.headers },
  });
  if (!response.ok) throw new Error(`API ${response.status}: ${path}`);
  return (await response.json()) as T;
}

export function fetchTwinScenarios(signal?: AbortSignal): Promise<ScenarioCatalog> {
  return readJson("/api/v1/twin/scenarios", signal);
}

export function createTwinRun(
  householdId: string,
  request: TwinRunRequest,
  signal?: AbortSignal,
): Promise<TwinRunStatus> {
  return readJson(`/api/v1/households/${encodeURIComponent(householdId)}/twin/runs`, signal, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
}

export function advanceTwinRun(
  householdId: string,
  runId: string,
  signal?: AbortSignal,
): Promise<TwinRunStatus> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/twin/runs/${encodeURIComponent(runId)}/advance`,
    signal,
    { method: "POST" },
  );
}

export function fetchTwinRun(
  householdId: string,
  runId: string,
  signal?: AbortSignal,
): Promise<TwinRunStatus> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/twin/runs/${encodeURIComponent(runId)}`,
    signal,
  );
}

export function cancelTwinRun(householdId: string, runId: string): Promise<TwinRunStatus> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/twin/runs/${encodeURIComponent(runId)}/cancel`,
    undefined,
    { method: "POST" },
  );
}

export function twinExportUrl(householdId: string, runId: string): string {
  return `${apiBase}/api/v1/households/${encodeURIComponent(householdId)}/twin/runs/${encodeURIComponent(runId)}/export`;
}

File: backend/app/schemas/twin.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Literal

from pydantic import BaseModel, BeforeValidator, Field, model_validator

from app.domain.enums import SimulationStatus
from app.schemas.records import Money, Ratio, SignedRatio, reject_binary_float

SignedMoney = Annotated[
    Decimal,
    BeforeValidator(reject_binary_float),
    Field(max_digits=20, decimal_places=2),
]


class AssetReturnOverride(BaseModel):
    expected_annual_return: Decimal | None = Field(
        default=None,
        gt=Decimal("-1"),
        lt=Decimal("1"),
    )
    annual_volatility: Decimal | None = Field(default=None, ge=0, le=Decimal("2"))


class AssumptionOverrides(BaseModel):
    inflation_rate: Decimal | None = Field(default=None, ge=0, le=Decimal("0.20"))
    income_growth_rate: Decimal | None = Field(
        default=None,
        ge=Decimal("-0.20"),
        le=Decimal("0.20"),
    )
    asset_assumptions: dict[str, AssetReturnOverride] = Field(default_factory=dict)
    correlation_matrix: list[list[Decimal]] | None = None


class ScenarioOverrides(BaseModel):
    unemployment_months: int | None = Field(default=None, ge=0, le=36)
    income_reduction_ratio: Ratio | None = None
    medical_shock_amount: Money | None = None
    education_overrun_amount: Money | None = None
    property_value_change_ratio: SignedRatio | None = None
    mortgage_rate_change: SignedRatio | None = None


class PlanAdjustments(BaseModel):
    primary_retirement_age: int | None = Field(default=None, ge=45, le=75)
    additional_monthly_savings: Money = Decimal("0.00")
    equity_ratio: Ratio | None = None
    liquidity_reallocation_amount: Money = Decimal("0.00")


class FamilyEventInput(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=160)
    start_month: int = Field(ge=1, le=720)
    duration_months: int = Field(default=1, ge=1, le=720)
    one_time_cost: Money = Decimal("0.00")
    monthly_income_loss: Money = Decimal("0.00")
    monthly_expense_increase: Money = Decimal("0.00")


class TwinRunRequest(BaseModel):
    analysis_date: date | None = None
    seed: int = Field(default=20260804, ge=0, le=2_147_483_647)
    path_count: int = Field(default=500, ge=100, le=5000)
    horizon_years: int = Field(default=30, ge=5, le=60)
    output_interval_months: Literal[1, 3, 6, 12] = 12
    scenario_codes: list[str] = Field(
        default_factory=lambda: ["unemployment_equity_down_30"],
        min_length=1,
        max_length=6,
    )
    scenario_overrides: ScenarioOverrides = Field(default_factory=ScenarioOverrides)
    plan_adjustments: PlanAdjustments = Field(default_factory=PlanAdjustments)
    assumption_overrides: AssumptionOverrides = Field(default_factory=AssumptionOverrides)
    family_events: list[FamilyEventInput] = Field(default_factory=list, max_length=12)

    @model_validator(mode="after")
    def validate_request(self) -> TwinRunRequest:
        if len(self.scenario_codes) != len(set(self.scenario_codes)):
            raise ValueError("压力场景代码不能重复")
        return self


class ScenarioDefinitionOut(BaseModel):
    code: str
    name: str
    category: str
    description: str
    parameters: dict[str, object]
    explanation: str
    scenario_version: str
    source_type: Literal["internal_demo"] = "internal_demo"
    is_composable: bool
    enabled: bool


class ScenarioCatalogResponse(BaseModel):
    scenario_version: str
    source_type: Literal["internal_demo"] = "internal_demo"
    source_summary: str
    scenario_count: int
    scenarios: list[ScenarioDefinitionOut]


class MemberState(BaseModel):
    member_id: str
    name: str
    relationship: str
    age_at_start: Decimal
    recorded_retirement_age: int | None


class GoalState(BaseModel):
    goal_id: str
    name: str
    goal_type: str
    due_month: int
    current_amount: Money
    prepared_amount: Money
    annual_cost_growth_rate: Decimal


class InitialTwinState(BaseModel):
    members: list[MemberState]
    annual_income: Money
    annual_expenses_excluding_debt_service: Money
    annual_essential_expenses_excluding_debt_service: Money
    monthly_compressible_expenses: Money
    asset_buckets: dict[str, Money]
    total_assets: Money
    total_liabilities: Money
    medical_coverage_available: Money
    medical_deductible: Money
    pension_annual_contributions: Money
    goals: list[GoalState]
    source_record_ids: list[str]
    counting_note: str


class ResolvedAssumptionSnapshot(BaseModel):
    seed: int
    path_count: int
    horizon_months: int
    time_step_months: int
    output_interval_months: int
    inflation_rate: Decimal
    income_growth_rate: Decimal
    asset_classes: list[str]
    asset_assumptions: dict[str, dict[str, Decimal]]
    correlation_matrix: list[list[Decimal]]
    property_assumption: dict[str, Decimal]
    pension_assumption: dict[str, Decimal]
    scenario_codes: list[str]
    scenario_parameters: dict[str, object]
    plan_adjustments: PlanAdjustments
    family_events: list[FamilyEventInput]
    parameter_hash: str
    source_type: Literal["internal_demo"] = "internal_demo"


class FanPoint(BaseModel):
    month: int
    date: date
    primary_age: Decimal
    p10: SignedMoney
    p25: SignedMoney
    p50: SignedMoney
    p75: SignedMoney
    p90: SignedMoney
    median_liquid_assets: Money
    median_long_term_assets: Money
    median_liabilities: Money


class GoalSimulationOutcome(BaseModel):
    goal_id: str
    name: str
    due_month: int
    required_amount: Money
    success_probability: Ratio
    failure_probability: Ratio
    median_shortfall: Money


class FailureTimeBucket(BaseModel):
    year: int
    path_count: int
    probability: Ratio
    most_common_goal: str | None


class WorstPath(BaseModel):
    path_id: int
    ending_net_worth: SignedMoney
    minimum_net_worth: SignedMoney
    first_failed_goal: str | None
    first_failure_month: int | None
    depletion_month: int | None
    forced_sale_amount: Money
    total_goal_shortfall: Money
    explanation: str


class DistributionValidation(BaseModel):
    all_values_finite: bool
    primary_income_paid_during_interruption_max: Money
    goal_spending_events_applied: int
    common_random_numbers: bool


class SimulationDistribution(BaseModel):
    label: str
    path_count: int
    horizon_months: int
    goal_success_probability: Ratio
    depletion_probability: Ratio
    forced_sale_probability: Ratio
    median_forced_sale_amount: Money
    ending_net_worth_median: SignedMoney
    ending_net_worth_p10: SignedMoney
    total_goal_shortfall_median: Money
    required_additional_monthly_savings: Money
    fan: list[FanPoint]
    goal_outcomes: list[GoalSimulationOutcome]
    failure_time_distribution: list[FailureTimeBucket]
    worst_paths: list[WorstPath]
    validation: DistributionValidation


class ScenarioImpact(BaseModel):
    scenario_code: str
    name: str
    emergency_support_months: Decimal
    first_failed_goal: str | None
    forced_sale_probability: Ratio
    median_forced_sale_amount: Money
    insurance_coverage_applied: Money
    remaining_medical_gap: Money
    retirement_delay_needed: bool
    suggested_retirement_delay_years: int
    monthly_compressible_expenses: Money
    required_additional_monthly_savings: Money
    baseline_goal_success_probability: Ratio
    scenario_goal_success_probability: Ratio
    success_probability_change: Decimal
    applicable: bool
    explanation: str


class TwinComparison(BaseModel):
    baseline_success_probability: Ratio
    original_stress_success_probability: Ratio
    optimized_stress_success_probability: Ratio
    stress_change: Decimal
    optimization_change: Decimal
    baseline_forced_sale_probability: Ratio
    original_forced_sale_probability: Ratio
    optimized_forced_sale_probability: Ratio
    liquidity_reallocation_amount: Money
    avoided_forced_sale_probability: Decimal
    stress_not_better_than_baseline: bool
    positive_override_explanation: str | None
    liquidity_explanation: str


class TwinResultMeta(BaseModel):
    run_id: str
    household_id: str
    household_code: str
    analysis_date: date
    data_as_of: date
    input_version: str
    rule_code: str
    rule_version: str
    formula_version: str
    engine_version: str
    result_version: str
    scenario_version: str
    calculation_source: Literal["deterministic_simulation_engine"] = (
        "deterministic_simulation_engine"
    )
    synthetic_data: bool
    currency: str


class TwinResult(BaseModel):
    meta: TwinResultMeta
    initial_state: InitialTwinState
    assumptions: ResolvedAssumptionSnapshot
    baseline: SimulationDistribution
    original_stress: SimulationDistribution
    optimized_stress: SimulationDistribution
    scenario_impacts: list[ScenarioImpact]
    comparison: TwinComparison
    limitations: list[str]
    counting_note: str


class TwinRunStatusResponse(BaseModel):
    run_id: str
    household_id: str
    status: SimulationStatus
    progress_percent: int = Field(ge=0, le=100)
    phase: str
    scenario_codes: list[str]
    path_count: int
    horizon_months: int
    seed: int
    input_version: str
    rule_version: str
    engine_version: str
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    cancel_requested: bool
    audit_event_id: str | None = None
    error_code: str | None
    result: TwinResult | None = None
